Exposed optional scrollController property in ReorderableListView

exts opened the pull request:
## Description

This exposes `scrollController` to the developer, allowing you to directly add a custom `ScrollController` without wrapping it with (or relying on) a PrimaryScrollController.

## Related Issues

#48219

## Tests

I added the following tests:

- Added check to see if custom widget was being set in `test\material\reorderable_list_test.dart`

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*

exts committed:
Exposed optional scrollController property to ReorderableListView constructor

exts commented:
@googlebot I signed it!

shihaohong commented on packages/flutter/lib/src/material/reorderable_list.dart:
```diff
@@ -262,7 +269,7 @@ class _ReorderableListContentState extends State<_ReorderableListContent> with T
   @override
   void didChangeDependencies() {
-    _scrollController = PrimaryScrollController.of(context) ?? ScrollController();
+    _scrollController = widget.scrollController ?? (PrimaryScrollController.of(context) ?? ScrollController());
```
The parentheses are probably unnecessary here:
```suggestion
    _scrollController = widget.scrollController ?? PrimaryScrollController.of(context) ?? ScrollController();
```

shihaohong commented on packages/flutter/test/material/reorderable_list_test.dart:
```diff
@@ -263,6 +263,26 @@ void main() {
 
+      testWidgets('Use custom ScrollController when scrollController is set', (WidgetTester tester) async {
+        final ScrollController customController = ScrollController();
+        final Widget reorderableList = ReorderableListView(
```
It's probably okay to compose the entire widget tree here like this, since `reorderableList` is not being reused in any way:
```dart
await tester.pumpWidget(
  MaterialApp(
    home: ReorderableListView(
      scrollController: customController,
      onReorder: (int oldIndex, int newIndex) {},
      children: const <Widget>[
        SizedBox(width: 100.0, height: 100.0, child: Text('C'), key: Key('C')),
        SizedBox(width: 100.0, height: 100.0, child: Text('B'), key: Key('B')),
        SizedBox(width: 100.0, height: 100.0, child: Text('A'), key: Key('A')),
      ],
    );
  ),
);
```

shihaohong commented on packages/flutter/test/material/reorderable_list_test.dart:
```diff
@@ -263,6 +263,26 @@ void main() {
+          find.byType(ReorderableListView),
+        );
+
+        expect(listView.scrollController, customController);
```
Instead of checking if `listView.scrollController == customController`, it's better to perform some operation on `customController` and see if that behavior correctly reflects on the UI. 

Let me know if you need any help with this!

exts replied:
@shihaohong yeah what would you recommend here?

shihaohong replied:
For example, passing in your own ScrollController and making sure that manipulating it results in visual changes. And also, when making interactions with the ReorderableListView (like dragging it around), the ScrollController's values are properly updated.

exts replied:
I'll give it a shot, I think there's an example in the existing test file for it that I could base it off.

shihaohong replied:
`tests/material/tabs_test.dart` should also have some tests that use `controller.jumpTo()`, then you could check the offsets of the children of the list view and see if they're expected

exts replied:
I added a jumpTo test (docs say this has no animations so I didn't call pump after setting jumpTo), but the tabs_test is for the TabController tests. Unrelated to this issue.

exts committed:
Removed redundant parenthesis & combined materialapp in test

exts committed:
Added a jumpTo check to make sure the custom scroller controller works

shihaohong commented on packages/flutter/test/material/reorderable_list_test.dart:
```diff
@@ -263,6 +263,31 @@ void main() {
+
+        listView.scrollController.jumpTo(20);
+
+        expect(customController.offset, 20);
```
Instead of checking the controller's offset value here, check the actual value position of the box relative to the list view. For example,
```dart
        // Update scroll controller and see if list view updates accordingly.
        customController.animateTo(
          40.0,
          duration: const Duration(milliseconds: 200),
          curve: Curves.linear,
        );
        await tester.pumpAndSettle();

        listViewTopLeft = tester.getTopLeft(
          find.byType(ReorderableListView),
        );
        firstBoxTopLeft = tester.getTopLeft(
          find.byKey(firstBox),
        );
        expect(firstBoxTopLeft.dy, listViewTopLeft.dy - 40.0);
```

shihaohong commented on packages/flutter/test/material/reorderable_list_test.dart:
```diff
@@ -263,6 +263,31 @@ void main() {
+            ),
+          )
+        );
+
```
We should check to see if the initial scroll offset was applied:
```dart
        // Check initial scroll offset of first list item relative to
        // the offset of the list view.
        Offset listViewTopLeft = tester.getTopLeft(
          find.byType(ReorderableListView),
        );
        Offset firstBoxTopLeft = tester.getTopLeft(
          find.byKey(firstBox),
        );
        expect(firstBoxTopLeft.dy, listViewTopLeft.dy - 10.0);
```

shihaohong commented on packages/flutter/test/material/reorderable_list_test.dart:
```diff
@@ -263,6 +263,31 @@ void main() {
+              onReorder: (int oldIndex, int newIndex) { },
+            ),
+          )
+        );
```
Wrap the `ReorderableListView` by a smaller box so that you can test scrolling the list view and seeing a measurable change. For example, if you constrain the list view to a height of 200, if you have three boxes in the list view at heights of 100.0 each, you can test scrolling the list view since the list view can be hidden:
```dart
await tester.pumpWidget(
  MaterialApp(
    home: Scaffold(
      body: SizedBox( // Constrains the height of the ListView to 200.0
        height: 200,
        child: ReorderableListView(
          scrollController: customController,
          children: const <Widget>[
            SizedBox(height: 100.0, child: Text('A'), key: firstBox),
            SizedBox(height: 100.0, child: Text('B'), key: secondBox),
            SizedBox(height: 100.0, child: Text('C'), key: thirdBox),
          ],
          onReorder: (int oldIndex, int newIndex) { },
        ),
      ),
    ),
  ),
);
```

shihaohong commented on packages/flutter/test/material/reorderable_list_test.dart:
```diff
@@ -263,6 +263,31 @@ void main() {
+
+        expect(customController.offset, 20);
+      });
+
```
You will also want to test dragging the list view and seeing if the controller's offset values update with changes on the screen. I did this one since it's a little odd. The first 20px of dragging seems to be ignored by the `ReorderableListView`:
```dart
        // Drag the UI to see if the scroll controller updates accordingly
        await tester.drag(
          find.text('B'),
          const Offset(0.0, -100.0),
        );
        listViewTopLeft = tester.getTopLeft(
          find.byType(ReorderableListView),
        );
        firstBoxTopLeft = tester.getTopLeft(
          find.byKey(firstBox),
        );
        // Initial scroll controller offset: 40.0
        // Drag UI by 100.0 upwards vertically
        // First 20.0 px always ignored, so scroll offset is only
        // shifted by 80.0.
        // Final offset: 40.0 + 80.0 = 120.0
        expect(customController.offset, 120.0);
```

exts replied:
You've taught me A LOT about the capabilities of flutter's testing library. Man lol, I need to go back over the docs a bit because a lot of this is new to me.

shihaohong reviewed (commented):
I have some comments to improve the test. Let me know if you have any questions!

exts committed:
Improved custom scrollcontroller behavioral test

shihaohong commented on packages/flutter/test/material/reorderable_list_test.dart:
```diff
@@ -263,6 +263,74 @@ void main() {
+                height: 200,
+                child: ReorderableListView(
+                  scrollController: customController,
+                  children: const <Widget>[
```
nit: prefer to have the `child` or `children` property to be the last property specified for `Widget`s. ie.
```dart
Row(
  // all properties here
  // ...
  children: [
    // children widgets here
  ],
)
```

shihaohong commented on packages/flutter/lib/src/material/reorderable_list.dart:
```diff
@@ -87,6 +88,9 @@ class ReorderableListView extends StatefulWidget {
   /// List [children] can only drag along this [Axis].
   final Axis scrollDirection;
 
+  /// Controls scrolls and measures scroll progress.
```
I would describe this a little more with regards to its use in a `ReorderableListView`. Use relevant info from [ScrollController](https://master-api.flutter.dev/flutter/widgets/ScrollController-class.html)'s docs, as well as describing common use cases for wanting to provide your own scroll controller.

shihaohong reviewed (commented):
I left a few more comments. 

This might just be my preference for reading code, but I prefer grouping the test code by its related setup and assertions instead of having a blank line between each line of code. For example, the following are related, so group them up like so: 
```dart
        // Drag the UI to see if the scroll controller updates accordingly
        await tester.drag(
          find.text('B'), 
          const Offset(0.0, -100.0),
        );
        listViewTopLeft = tester.getTopLeft(
          find.byType(ReorderableListView),
        );
        firstBoxTopLeft = tester.getTopLeft(
          find.byKey(firstBox),
        );
        // Initial scroll controller offset: 40.0
        // Drag UI by 100.0 upwards vertically
        // First 20.0 px always ignored, so scroll offset is only
        // shifted by 80.0.
        // Final offset: 40.0 + 80.0 = 120.0
        expect(customController.offset, 120.0);
```

shihaohong commented:
Also, the analyzer seems to be complaining about trailing spaces. Be sure to run `flutter analyze` in the `flutter/packages/flutter` repo to ensure that the analyzer passes without issues before every push to GitHub

shihaohong commented:
Unrelated tests also seem to be failing, which means you might have to rebase your branch against `upstream/master`.

exts committed:
Improved documentation and trimmed whitespace

exts commented:
Alright, I rebased with upstream master. I reran flutter analyze **(no issues)**. I reran all tests **(all passed)** and I cleaned up the description & whitespace. Should be good to go.

shihaohong commented:
It seems there are still a few trailing spaces in your changes :)

exts commented:
lol can you point them out because I don't see it. Are you talking about 289 and 305? I think that would make it even less readable compared to every other test in that file

shihaohong commented on packages/flutter/lib/src/material/reorderable_list.dart:
```diff
@@ -87,6 +88,10 @@ class ReorderableListView extends StatefulWidget {
   final Axis scrollDirection;
 
+  /// Creates a ScrollPosition to manage and determine which portion
+  /// of the content is visible in a scroll view.
```
I would further elaborate here on how it is used for `ReorderableListView`:
```dart
/// Creates a [ScrollPosition] to manage and determine which portion
/// of the content is visible in the scroll view.
///
/// This can be used in many ways, such as setting an initial scroll offset, 
/// (via [ScrollController.initialScrollOffset]), reading the current scroll position 
/// (via [ScrollController.offset]), or changing it (via [ScrollController.jumpTo] or 
/// [ScrollController.animateTo]).
```

shihaohong reviewed (approved):
LGTM other than the API doc feedback and the flutter analyzer issues that need to be addressed

shihaohong commented:
Trailing spaces are spaces that occur after any code. In the current failure, it's saying:
```
/tmp/flutter sdk/packages/flutter/test/material/reorderable_list_test.dart:293: trailing U+0020 space character
/tmp/flutter sdk/packages/flutter/test/material/reorderable_list_test.dart:294: trailing U+0020 space character
/tmp/flutter sdk/packages/flutter/test/material/reorderable_list_test.dart:308: trailing U+0020 space character
```
Let me highlight them in a code review

shihaohong commented on packages/flutter/test/material/reorderable_list_test.dart:
```diff
@@ -263,6 +263,65 @@ void main() {
+        // Check initial scroll offset of first list item relative to
+        // the offset of the list view.
+        customController.animateTo(
+          40.0, 
```
```suggestion
          40.0,
```

shihaohong commented on packages/flutter/test/material/reorderable_list_test.dart:
```diff
@@ -263,6 +263,65 @@ void main() {
+        // the offset of the list view.
+        customController.animateTo(
+          40.0, 
+          duration: const Duration(milliseconds: 200), 
```
```suggestion
          duration: const Duration(milliseconds: 200),
```

shihaohong commented on packages/flutter/test/material/reorderable_list_test.dart:
```diff
@@ -263,6 +263,65 @@ void main() {
+
+        // Drag the UI to see if the scroll controller updates accordingly
+        await tester.drag(
+          find.text('B'), 
```
```suggestion
          find.text('B'),
```

shihaohong reviewed (commented):
Remove trailing whitespaces

exts committed:
Improved doc information for the scrollController property

exts committed:
removed trailing spaces

shihaohong commented on packages/flutter/lib/src/material/reorderable_list.dart:
```diff
@@ -87,6 +88,15 @@ class ReorderableListView extends StatefulWidget {
+  /// Creates a [ScrollPosition] to manage and determine which portion
+  /// of the content is visible in the scroll view.
+  ///
+  /// This can be used in many ways, such as setting an initial scroll offset, 
```
```suggestion
  /// This can be used in many ways, such as setting an initial scroll offset,
```

shihaohong commented on packages/flutter/lib/src/material/reorderable_list.dart:
```diff
@@ -87,6 +88,15 @@ class ReorderableListView extends StatefulWidget {
+  /// of the content is visible in the scroll view.
+  ///
+  /// This can be used in many ways, such as setting an initial scroll offset, 
+  /// (via [ScrollController.initialScrollOffset]), reading the current scroll position 
```
```suggestion
  /// (via [ScrollController.initialScrollOffset]), reading the current scroll position
```

shihaohong commented on packages/flutter/lib/src/material/reorderable_list.dart:
```diff
@@ -87,6 +88,15 @@ class ReorderableListView extends StatefulWidget {
+  ///
+  /// This can be used in many ways, such as setting an initial scroll offset, 
+  /// (via [ScrollController.initialScrollOffset]), reading the current scroll position 
+  /// (via [ScrollController.offset]), or changing it (via [ScrollController.jumpTo] or 
```
```suggestion
  /// (via [ScrollController.offset]), or changing it (via [ScrollController.jumpTo] or
```

exts committed:
Removed whitespace from docblock
